modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/ModifyNode.java

```diff
@@ -465,9 +479,25 @@ private boolean hasUpsertSemantic(int rowSize) {
 
             assert !columnDescriptor.virtual() : "Virtual column can't be updated";
 
-            mapping[columnDescriptor.logicalIndex()] = columnCount + i;
+            mapping[columnDescriptor.logicalIndex()] = columnCount + i + mergeOffset;
         }
 
         return mapping;
     }
+
+    private RowSchema getMappedRowSchema(RowFactory<RowT> inputRowFactory, int[] mapping) {
+        RowSchema schema;
+        if (mapping != null) {
+            List<TypeSpec> fields = inputRowFactory.rowSchema().fields();
+            Builder builder = RowSchema.builder();
+            for (int i : mapping) {
+                TypeSpec typeSpec = fields.get(i);
+                builder.addField(typeSpec);
+            }
+            schema = builder.build();
+        } else {
+            schema = inputRowFactory.rowSchema();
+        }
+        return schema;
+    }
```

Comment:
It is not the only class where this method is useful.
Can it be a static method in RowSchema for example?

Reply:
I meant this
```
class RowSchema {
    static RowSchema map(RowSchema schema, int[] mapping) {
        List<TypeSpec> fields = schema.fields();
        Builder builder = builder();

        for (int i : mapping) {
            TypeSpec typeSpec = fields.get(i);
            builder.addField(typeSpec);
        }

        return builder.build();
    }
}
```